backend/src/stwo/prover.rs

```diff
@@ -1,32 +1,144 @@
 use powdr_ast::analyzed::Analyzed;
+use serde::de::DeserializeOwned;
+use serde::ser::Serialize;
 use std::io;
+use std::marker::PhantomData;
 use std::sync::Arc;
 
+use crate::stwo::circuit_builder::gen_stwo_circuit_trace;
+use crate::stwo::circuit_builder::PowdrEval;
+
+use super::circuit_builder::PowdrComponent;
+
+use stwo_prover::constraint_framework::TraceLocationAllocator;
+use stwo_prover::core::prover::StarkProof;
+
 use powdr_number::FieldElement;
+use stwo_prover::core::air::Component;
+use stwo_prover::core::air::ComponentProver;
+use stwo_prover::core::backend::Backend;
+use stwo_prover::core::backend::BackendForChannel;
+use stwo_prover::core::channel::Channel;
+use stwo_prover::core::channel::MerkleChannel;
+use stwo_prover::core::fields::m31::M31;
+use stwo_prover::core::fri::FriConfig;
+use stwo_prover::core::pcs::{CommitmentSchemeProver, CommitmentSchemeVerifier, PcsConfig};
+use stwo_prover::core::poly::circle::CanonicCoset;
+
+const FRI_LOG_BLOWUP: usize = 1;
+const FRI_NUM_QUERIES: usize = 100;
+const FRI_PROOF_OF_WORK_BITS: usize = 16;
+const LOG_LAST_LAYER_DEGREE_BOUND: usize = 0;
 
 #[allow(unused_variables)]
-pub struct StwoProver<F> {
-    _analyzed: Arc<Analyzed<F>>,
-    _fixed: Arc<Vec<(String, Vec<F>)>>,
+pub struct StwoProver<T, B: Backend + Send, MC: MerkleChannel, C: Channel> {
+    pub analyzed: Arc<Analyzed<T>>,
+    _fixed: Arc<Vec<(String, Vec<T>)>>,
     /// Proving key placeholder
     _proving_key: Option<()>,
     /// Verifying key placeholder
     _verifying_key: Option<()>,
+    pub pcs_config: PcsConfig,
+    _channel_marker: PhantomData<C>,
+    _channel_marker_backend: PhantomData<B>,
+    _channel_marker_merkelchannel: PhantomData<MC>,
 }
 
-impl<F: FieldElement> StwoProver<F> {
+impl<'a, F: FieldElement, B: Backend + Send, MC: MerkleChannel + Send, C: Channel + Send>
+    StwoProver<F, B, MC, C>
+where
+    B: Backend + BackendForChannel<MC>, // Ensure B implements BackendForChannel<MC>
+    MC: MerkleChannel,
+    C: Channel,
+    MC::H: DeserializeOwned + Serialize,
+    PowdrComponent<'a, F>: ComponentProver<B>,
+{
     #[allow(dead_code)]
     #[allow(unused_variables)]
     pub fn new(
-        _analyzed: Arc<Analyzed<F>>,
+        analyzed: Arc<Analyzed<F>>,
         _fixed: Arc<Vec<(String, Vec<F>)>>,
-        setup: Option<&mut dyn io::Read>,
+        //  setup: Option<&mut dyn io::Read>,
     ) -> Result<Self, io::Error> {
+        let config = PcsConfig {
+            pow_bits: FRI_PROOF_OF_WORK_BITS as u32,
+            fri_config: FriConfig::new(
+                LOG_LAST_LAYER_DEGREE_BOUND as u32,
+                FRI_LOG_BLOWUP as u32,
+                FRI_NUM_QUERIES,
+            ),
+        };
```

Comment:
I'm not convinced this should be stored in `StwoProver`: it is constant and cannot be set in the constructor. Instead we could have a function which returns it.

Reply:
```
fn get_config() -> PcsConfig {
    ...
}
```

Reply:
function added